### PR TITLE
[CP]Update pod label based on the master election(#908)

### PR DESCRIPTION
**Pull request by zhengxiexie**
To ensure the webhook targets the active endpoint in HA mode, we should label the pods accordingly and configure the webhook service to select these labeled pods.

### CODE REVIEW

File: build/yaml/webhook/service.yaml [hunk 1]
```diff
@@ -18,3 +18,4 @@ spec:
       targetPort: 9981
   selector:
     component: nsx-ncp
+    role: master
```
Comment (dantingl) [inline build/yaml/webhook/service.yaml:21]: Is there any way to identify it is nsx-operator master, as it could be a different Pod for ncp master.

File: cmd/main.go [hunk 2]
```diff
@@ -261,6 +271,12 @@ func startServiceController(mgr manager.Manager, nsxClient *nsx.Client) {
 	if cf.EnableAntreaNSXInterworking {
 		StartNSXServiceAccountController(mgr, commonService)
 	}
+
+	// Update pod labels to determine if this pod is the master
+	err := updatePodLabels(mgr)
```
Comment (dantingl) [inline cmd/main.go:276]: Will nsx-operator restart if updatePodLabels fails?

File: cmd/main.go [hunk 3]
```diff
@@ -395,3 +411,31 @@ func refreshCertPeriodically() {
...
+	}
+
+	// Iterate over the pods and update the role labels
+	for _, po := range podList.Items {
+		targetRole := roleStandby
+		if po.Name == nsxOperatorPodName {
+			targetRole = roleMaster
+		}
+
+		if po.Labels["role"] != targetRole {
```
Comment (dantingl) [inline cmd/main.go:431]: Can we only add label for master, no need to add standby label, right?

Comment (dantingl) [reply cmd/main.go:431]: I guess in that case, we need to remove old master label when there is a leader election again.

File: build/yaml/webhook/service.yaml (hunk 1 above)
Comment (zhengxiexie) [reply build/yaml/webhook/service.yaml:21]: I suppose it is the clearest, any problem for a different Pod for ncp master?

File: cmd/main.go (hunk 2 above)
Comment (zhengxiexie) [reply cmd/main.go:276]: I added panic, yes

File: cmd/main.go (hunk 3 above)
Comment (zhengxiexie) [reply cmd/main.go:432]: yes, it will be removed.  I think `standby`  is helpful for debugging.

File: test/e2e/nsx_networkinfo_test.go
Comment (wenqiq) [inline test/e2e/nsx_networkinfo_test.go:1]: It seems that this file is not the latest version. Should rebase? And the changes here are not related to this PR?

File: cmd/main.go [hunk 4]
```diff
@@ -261,6 +271,13 @@ func startServiceController(mgr manager.Manager, nsxClient *nsx.Client) {
 	if cf.EnableAntreaNSXInterworking {
 		StartNSXServiceAccountController(mgr, commonService)
 	}
+
+	// Update pod labels to determine if this pod is the master
+	err := updatePodLabels(mgr)
+	if err != nil {
+		log.Error(err, "Failed to update pod labels")
```
Comment (wenqiq) [inline cmd/main.go:278]: Proper nouns are better capitalized—for example, `Pod`, `Namespace`, etc.

File: cmd/main.go [hunk 5]
```diff
@@ -395,3 +412,32 @@ func refreshCertPeriodically() {
...
 	}
 }
+
+// updatePodLabels updates the role label of pods based on the master election.
+func updatePodLabels(mgr manager.Manager) error {
+	c := mgr.GetClient()
+	// Fetch all pods in the given namespace
+	podList := &corev1.PodList{}
+	if err := c.List(context.TODO(), podList, &client.ListOptions{Namespace: nsxOperatorNamespace}); err != nil {
+		return fmt.Errorf("failed to list pods in namespace %s: %w", nsxOperatorNamespace, err)
```
Comment (wenqiq) [inline cmd/main.go:422]: ditto

File: cmd/main.go [hunk 6]
```diff
@@ -395,3 +412,32 @@ func refreshCertPeriodically() {
...
+		targetRole := roleStandby
+		if po.Name == nsxOperatorPodName {
+			targetRole = roleMaster
+		}
+
+		if po.Labels["role"] != targetRole {
+			patch := client.MergeFrom(po.DeepCopy())
+			po.Labels["role"] = targetRole
+			if err := c.Patch(context.TODO(), &po, patch); err != nil {
+				return fmt.Errorf("failed to update labels for pod %s: %w", po.Name, err)
```
Comment (wenqiq) [inline cmd/main.go:436]: ditto

File: cmd/main.go [hunk 7]
```diff
@@ -395,3 +412,32 @@ func refreshCertPeriodically() {
...
+			}
+			log.Info("Updated pod labels", "pod", po.Name, "labels", po.Labels)
```
Comment (wenqiq) [inline cmd/main.go:438]: ditto

File: test/e2e/nsx_networkinfo_test.go
Comment (zhengxiexie) [reply test/e2e/nsx_networkinfo_test.go:1]: fixed

File: cmd/main.go (hunk 4 above)
Comment (zhengxiexie) [reply cmd/main.go:278]: ok

File: build/yaml/webhook/service.yaml (hunk 1 above)
Comment (zhengxiexie) [reply build/yaml/webhook/service.yaml:22]: Since all tests passed, added a comment to this label.

**Review by dantingl — DISMISSED** [review]

**Review by dantingl — APPROVED** [review]

**Review by yanjunz97 — APPROVED** [review]
